internal/util/fetch/fetch.go

```diff
@@ -94,21 +94,62 @@ func (c Command) validate(kf *kptfilev1.KptFile) error {
 	return nil
 }
 
+type Cloner struct {
+	// repoSpec spec to clone
+	repoSpec *git.RepoSpec
+
+	// cachedRepos
+	cachedRepo map[git.RepoSpec]*gitutil.GitUpstreamRepo
+
+	// fetchedRefs keeps track of refs already fetched from remote
+	fetchedRefs map[string]bool
+}
+
+type NewClonerOption func(*Cloner)
+
+func WithCachedRepo(r map[git.RepoSpec]*gitutil.GitUpstreamRepo) NewClonerOption {
+	return func(c *Cloner) {
+		c.cachedRepo = r
+	}
+}
+
+func WithAlreadyFetchedRefs(a map[string]bool) NewClonerOption {
+	return func(c *Cloner) {
+		c.fetchedRefs = a
+	}
+}
+
+func NewCloner(r *git.RepoSpec, opts ...NewClonerOption) *Cloner {
+	c := &Cloner{
+		repoSpec: r,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if c.cachedRepo == nil {
+		c.cachedRepo = make(map[git.RepoSpec]*gitutil.GitUpstreamRepo)
+	}
+	if c.fetchedRefs == nil {
+		c.fetchedRefs = map[string]bool{}
+	}
+	return c
+}
+
 // cloneAndCopy fetches the provided repo and copies the content into the
 // directory specified by dest. The provided name is set as `metadata.name`
 // of the Kptfile of the package.
-func cloneAndCopy(ctx context.Context, r *git.RepoSpec, dest string) error {
+func (c *Cloner) cloneAndCopy(ctx context.Context, dest string) error {
 	const op errors.Op = "fetch.cloneAndCopy"
 	pr := printer.FromContextOrDie(ctx)
 
-	err := ClonerUsingGitExec(ctx, r)
+	err := c.ClonerUsingGitExec(ctx)
 	if err != nil {
 		return errors.E(op, errors.Git, types.UniquePath(dest), err)
 	}
-	defer os.RemoveAll(r.Dir)
+	defer os.RemoveAll(c.repoSpec.Dir)
```

Comment:
what does this means wrt caching (do the fetchedRefs need to be cleaned up ?)

Reply:
Missed this, we should clean up